File: bindings/python/examples/compound.py
"""Coherent plane-wave compounding.

Plane-wave compounding transmits at several steering angles and
combines the resulting images to improve lateral resolution and
reduce artifacts. ffdas.das handles this as follows: the
sequence dimension of the rf data holds per-angle channel data,
and the matching offsets array provides the per-angle transmit
delay at each voxel. DAS sums over the sequence dimension,
producing a coherently compounded image.
"""

import cupy as cp
import matplotlib.pyplot as plt

import ffdas


cp.random.seed(0)

sound_speed = 1540.0
center_freq = 3.08e6
sampling_freq = center_freq
n_samples = 512

pitch = 500e-6  # 1 wavelength
el = (cp.arange(32, dtype=cp.float32) - 15.5) * pitch
ex, ey = cp.meshgrid(el, el, indexing="ij")  # type: ignore
channel_pos = cp.column_stack([ex.ravel(), ey.ravel(), cp.zeros_like(ex).ravel()])
n_channels = channel_pos.shape[0]

xmin, xmax = -0.008, 0.008
ymin, ymax = -0.008, 0.008
zmin, zmax = 0.010, 0.026

# 2D angle grid: 3 angles along x times 3 along y = 9 compound angles
angle_range = cp.linspace(-10, 10, 3, dtype=cp.float32) * cp.pi / 180  # type: ignore
angle_x, angle_y = cp.meshgrid(angle_range, angle_range, indexing="ij")  # type: ignore
angle_x = angle_x.ravel()
angle_y = angle_y.ravel()
n_angles = angle_x.shape[0]


# simulate rf data for each steering angle.
# a plane wave at angle (theta_x, theta_y) steers by delaying element i
# by (sin(theta_x) * element_x + sin(theta_y) * element_y) / c


def trefoil(t):
    return cp.stack(
        [
            (cp.sin(t) + 2 * cp.sin(2 * t)) * 0.1 * (xmax - xmin),
            -cp.sin(3 * t) * 0.15 * (ymax - ymin),
            (cp.cos(t) - 2 * cp.cos(2 * t)) * 0.1 * (zmax - zmin) + 0.5 * (zmin + zmax),
        ],
        axis=-1,
    )


batch_size = 16
n_scatterers = 16384
n_knot = 8192

t = 2 * cp.pi * cp.random.rand(n_knot, dtype=cp.float32)
knot_pos = trefoil(t) + 0.00015 * cp.random.randn(n_knot, 3, dtype=cp.float32)
background_pos = cp.random.uniform(
    [xmin, ymin, zmin],  # type: ignore
    [xmax, ymax, zmax],  # type: ignore
    size=(n_scatterers - n_knot, 3),
    dtype=cp.float32,  # type: ignore
)

scatter_pos = cp.concat([knot_pos, background_pos], axis=0)
scatter_values = cp.concat(
    [
        cp.random.rand(batch_size, n_knot, dtype=cp.float32),
        cp.random.rand(batch_size, n_scatterers - n_knot, dtype=cp.float32),
    ],
    axis=1,
)

freqs = cp.fft.fftfreq(n_samples, d=1.0 / sampling_freq).astype(cp.float32) + center_freq
wavenums = (-2 * cp.pi * freqs / sound_speed).astype(cp.float32)
sigma_f = 0.6 * center_freq / (2 * cp.sqrt(2 * cp.log(2)))
pulse = cp.exp(-0.5 * ((freqs - center_freq) / sigma_f) ** 2).astype(cp.complex64)

rf = cp.empty((batch_size, n_channels, n_angles, n_samples), dtype=cp.complex64)
for i in range(n_angles):
    delay = (
        float(cp.sin(angle_x[i])) * channel_pos[:, 0] + float(cp.sin(angle_y[i])) * channel_pos[:, 1]
    ) / sound_speed
    tx_signal = pulse * cp.exp(-2j * cp.pi * freqs * delay[:, None])
    tx = ffdas.greens(channel_pos, wavenums, tx_signal, scatter_pos)
    rx = ffdas.greens(scatter_pos, wavenums, tx * scatter_values[:, :, None], channel_pos)
    rf[:, :, i] = cp.fft.ifft(rx, axis=-1).conj()

# reconstruction grid
nz, ny, nx = 64, 64, 64
x = cp.linspace(xmin, xmax, nx, dtype=cp.float32)
y = cp.linspace(ymin, ymax, ny, dtype=cp.float32)
z = cp.linspace(zmin, zmax, nz, dtype=cp.float32)
zz, yy, xx = cp.meshgrid(z, y, x, indexing="ij")  # type: ignore
voxel_pos = cp.stack([xx, yy, zz], axis=-1)

ks = sampling_freq / sound_speed

# per-angle transmit offsets: one-way plane-wave delay at angles (theta_x, theta_y)
# is (x*sin(theta_x) + y*sin(theta_y) + z*cos(theta_x)*cos(theta_y)) * fs/c
sin_ax = cp.sin(angle_x)[:, None, None, None]
sin_ay = cp.sin(angle_y)[:, None, None, None]
cos_ax = cp.cos(angle_x)[:, None, None, None]
cos_ay = cp.cos(angle_y)[:, None, None, None]
offsets = (xx * sin_ax + yy * sin_ay + zz * cos_ax * cos_ay) * ks  # (n_angles, nz, ny, nx)
weights = cp.ones_like(offsets)

wavenum = -2 * cp.pi * center_freq / sampling_freq

# compounded: das sums over the sequence dimension (all angles at once)
with ffdas.utils.Timer() as t:
    compounded = ffdas.das(
        rf,
        channel_pos * ks,
        voxel_pos * ks,
        offsets,
        weights,
        wavenum=wavenum,
    )
print(f"compounded ({n_angles} angles): {t.elapsed_ms():.1f} ms")

# single angle at normal incidence for comparison
mid = n_angles // 2
with ffdas.utils.Timer() as t:
    single = ffdas.das(
        rf[:, :, mid : mid + 1, :],
        channel_pos * ks,
        voxel_pos * ks,
        offsets[mid : mid + 1],
        weights[mid : mid + 1],
        wavenum=wavenum,
    )
print(f"single angle: {t.elapsed_ms():.1f} ms")


mag_single = cp.abs(single[0])
mag_compounded = cp.abs(compounded[0])

fig, axes = plt.subplots(2, 2, figsize=(10, 10), sharey="row", sharex="col")

extent_xz = [float(x[0]) * 1e3, float(x[-1]) * 1e3, float(z[-1]) * 1e3, float(z[0]) * 1e3]
extent_yz = [float(y[0]) * 1e3, float(y[-1]) * 1e3, float(z[-1]) * 1e3, float(z[0]) * 1e3]

mip = mag_single.max(axis=1)
db = 20 * cp.log10(mip / mip.max() + 1e-10)
axes[0, 0].imshow(cp.asnumpy(db), extent=extent_xz, cmap="gray", vmin=-24, vmax=0, aspect="equal")
axes[0, 0].set_xlabel("x [mm]")
axes[0, 0].set_ylabel("z [mm]")
axes[0, 0].set_title("single angle — xz")

mip = mag_single.max(axis=2)
db = 20 * cp.log10(mip / mip.max() + 1e-10)
axes[0, 1].imshow(cp.asnumpy(db), extent=extent_yz, cmap="gray", vmin=-24, vmax=0, aspect="equal")
axes[0, 1].set_xlabel("y [mm]")
axes[0, 1].set_title("single angle — yz")

mip = mag_compounded.max(axis=1)
db = 20 * cp.log10(mip / mip.max() + 1e-10)
axes[1, 0].imshow(cp.asnumpy(db), extent=extent_xz, cmap="gray", vmin=-24, vmax=0, aspect="equal")
axes[1, 0].set_xlabel("x [mm]")
axes[1, 0].set_ylabel("z [mm]")
axes[1, 0].set_title(f"compounded ({n_angles} angles) — xz")

mip = mag_compounded.max(axis=2)
db = 20 * cp.log10(mip / mip.max() + 1e-10)
axes[1, 1].imshow(cp.asnumpy(db), extent=extent_yz, cmap="gray", vmin=-24, vmax=0, aspect="equal")
axes[1, 1].set_xlabel("y [mm]")
axes[1, 1].set_title(f"compounded ({n_angles} angles) — yz")

plt.tight_layout()
plt.savefig("compound.png", dpi=150)

File: bindings/matlab/examples/compound.m
% Coherent plane-wave compounding.
%
% Plane-wave compounding transmits at several steering angles and
% combines the resulting images to improve lateral resolution and
% reduce artifacts. ffdas.das handles this as follows: the
% sequence dimension of the rf data holds per-angle channel data,
% and the matching offsets array provides the per-angle transmit
% delay at each voxel. DAS sums over the sequence dimension,
% producing a coherently compounded image.

rng(0);

sound_speed = 1540.0;
center_freq = 3e6;
sampling_freq = center_freq;
n_samples = 512;

pitch = 500e-6;
el = gpuArray(single((0:31) - 15.5) * pitch);
[ex, ey] = ndgrid(el, el);
n_channels = 1024;
channel_pos = gpuArray(single([ex(:)'; ey(:)'; zeros(1, n_channels)]));

angle_range = gpuArray(single(linspace(-10, 10, 3))) * (pi / 180);
[angle_x, angle_y] = ndgrid(angle_range, angle_range);
angle_x = angle_x(:);
angle_y = angle_y(:);
n_angles = numel(angle_x);


% simulate rf data for each steering angle

batch_size = 32;
n_scatterers = 4096;
t = 2 * pi * rand(1, n_scatterers, "single", "gpuArray");
scatter_pos = trefoil(t) + randn(3, n_scatterers, "single", "gpuArray") * 0.00015;

k_fft = gpuArray(single([0:n_samples/2-1, -n_samples/2:-1]'));
freqs = k_fft * sampling_freq / n_samples + center_freq;
wavenums = single(-2 * pi) * freqs / sound_speed;
sigma_f = 0.6 * center_freq / (2 * sqrt(2 * log(2)));
pulse = complex(exp(-0.5 * ((freqs - center_freq) / sigma_f).^2));
scatter_values = rand(1, n_scatterers, batch_size, "single", "gpuArray");

rf = zeros(n_samples, n_angles, n_channels, batch_size, "single", "gpuArray");
for i = 1:n_angles
    delay = (sin(angle_x(i)) * channel_pos(1,:) ...
           + sin(angle_y(i)) * channel_pos(2,:)) / sound_speed;
    tx_signal = pulse .* exp(-2j * pi * freqs .* delay);
    tx = ffdas.greens(channel_pos, wavenums, tx_signal, scatter_pos);
    rx = ffdas.greens(scatter_pos, wavenums, tx .* scatter_values, channel_pos);
    rf_angle = conj(ifft(rx, [], 1));
    rf(:, i, :, :) = reshape(rf_angle, n_samples, 1, [], batch_size);
end


% reconstruction grid
nz = 64; ny = 64; nx = 64;
x = gpuArray(single(linspace(-0.007, 0.007, nx)));
y = gpuArray(single(linspace(-0.007, 0.007, ny)));
z = gpuArray(single(linspace(0.003, 0.017, nz)));
[xx, yy, zz] = ndgrid(x, y, z);
voxel_pos = zeros(3, nx, ny, nz, "single", "gpuArray");
voxel_pos(1,:,:,:) = xx;
voxel_pos(2,:,:,:) = yy;
voxel_pos(3,:,:,:) = zz;

ks = sampling_freq / sound_speed;

% per-angle transmit offsets: plane-wave delay at (theta_x, theta_y)
sin_ax = reshape(sin(angle_x), 1, 1, 1, []);
sin_ay = reshape(sin(angle_y), 1, 1, 1, []);
cos_ax = reshape(cos(angle_x), 1, 1, 1, []);
cos_ay = reshape(cos(angle_y), 1, 1, 1, []);
offsets = (xx .* sin_ax + yy .* sin_ay + zz .* cos_ax .* cos_ay) * ks;
weights = ones(size(offsets), "single", "gpuArray");

srcdir = zeros(4, n_channels, "single", "gpuArray");
srcdir(3,:) = 1.0;
srcdir(4,:) = 0.5;

wavenum = single(-2 * pi * center_freq / sampling_freq);

% compounded: das sums over the sequence dimension (all angles at once)
timer = ffdas.utils.Timer();
timer.start();
compounded = ffdas.das( ...
    rf, channel_pos * ks, voxel_pos * ks, offsets, weights, srcdir, wavenum);
timer.stop();
fprintf("compounded (%d angles): %.1f ms\n", n_angles, timer.elapsed_ms());

% single angle at normal incidence for comparison
mid = ceil(n_angles / 2);
timer2 = ffdas.utils.Timer();
timer2.start();
single_img = ffdas.das( ...
    rf(:, mid:mid, :, :), ...
    channel_pos * ks, voxel_pos * ks, ...
    offsets(:,:,:,mid:mid), weights(:,:,:,mid:mid), ...
    srcdir, wavenum);
timer2.stop();
fprintf("single angle: %.1f ms\n", timer2.elapsed_ms());


mag_single = abs(single_img(:,:,:,1));
mag_compounded = abs(compounded(:,:,:,1));

x_mm = gather(x) * 1e3;
y_mm = gather(y) * 1e3;
z_mm = gather(z) * 1e3;

figure;
tiledlayout(2, 2);

mip = squeeze(max(mag_single, [], 2));
db = 20 * log10(mip / max(mip(:)) + 1e-10);
nexttile;
imagesc(x_mm, z_mm, gather(db)');
colormap("gray"); clim([-24 0]); axis image;
xlabel("x [mm]"); ylabel("z [mm]");
title("single angle — xz");

mip = squeeze(max(mag_single, [], 1));
db = 20 * log10(mip / max(mip(:)) + 1e-10);
nexttile;
imagesc(y_mm, z_mm, gather(db)');
colormap("gray"); clim([-24 0]); axis image;
xlabel("y [mm]");
title("single angle — yz");

mip = squeeze(max(mag_compounded, [], 2));
db = 20 * log10(mip / max(mip(:)) + 1e-10);
nexttile;
imagesc(x_mm, z_mm, gather(db)');
colormap("gray"); clim([-24 0]); axis image;
xlabel("x [mm]"); ylabel("z [mm]");
title(sprintf("compounded (%d angles) — xz", n_angles));

mip = squeeze(max(mag_compounded, [], 1));
db = 20 * log10(mip / max(mip(:)) + 1e-10);
nexttile;
imagesc(y_mm, z_mm, gather(db)');
colormap("gray"); clim([-24 0]); axis image;
xlabel("y [mm]");
title(sprintf("compounded (%d angles) — yz", n_angles));

exportgraphics(gcf, "compound.png", Resolution=150);


function pos = trefoil(t)
    pos = cat(1, ...
        (sin(t) + 2 * sin(2*t)) * 0.0015, ...
        -sin(3*t) * 0.002, ...
        (cos(t) - 2 * cos(2*t)) * 0.0015 + 0.01);
end
